### PR TITLE
Fix MemoryAsync to keep a weakref to stream

### PR DESCRIPTION
**Pull request by kmaehashi**
Currently MemoryAsync is holding a pointer to the stream, but this is unsafe because:
* In general it is prohibited to pass destoryed ptr to CUDA APIs (I sometimes experienced SegV).
* The stream ptr may be reused.  (e.g., `cudaStreamCreate(&x); cudaStreamDestroy(x); cudaStreamCreate(&y)  // x != y is not guaranteed`)

ref: #4592

### CODE REVIEW

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by kmaehashi** [comment]
cc/ @leofang

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -135,48 +135,33 @@ cdef class MemoryAsync(BaseMemory):
...
+    def __init__(self, size_t size, stream):
         self.size = size
         self.device_id = device.get_device_id()
         # The stream is allowed to be destroyed before the memory is freed, so
-        # we don't need to hold a reference to the stream.
-        self.stream = stream
-        is_async_alloc_supported(self.device_id)
+        # we don't need to hold a strong reference to the stream.
+        self.stream_ref = weakref.ref(stream)
+        self.ptr = 0
```
Comment (leofang) [inline cupy/cuda/memory.pyx:150]: Just a nitpick: cdef class attribute is zero-initialized, so we don't have to add this line.
```suggestion
```

**Review by leofang — DISMISSED** [review]
LGTM! Just left a nitpick. 

It is weird that I cannot reproduce any errors or segfaults on my side... I thought the ok_errors are enough to handle the stream lifetime issue 🤔

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Review by takagi — APPROVED** [review]

**Pull request merged**